ipmanager/hetznerFloatingIpConfigurer.go

```diff
@@ -0,0 +1,193 @@
+
+import (
+	"context"
+	"errors"
+	"log"
+	"time"
+
+	"github.com/cybertec-postgresql/vip-manager/vipconfig"
+
+	"github.com/hetznercloud/hcloud-go/hcloud"
+)
+
+type HetznerFloatingIpConfigurer struct {
+	*IPConfiguration
+	client       *hcloud.Client
+	server       *hcloud.Server
+	floatingIpId string
+	cachedState  int
+	lastAPICheck time.Time
+	verbose      bool
+}
+
+func newHetznerFloatingIpConfigurer(config *vipconfig.Config, ipConfig *IPConfiguration) (*HetznerFloatingIpConfigurer, error) {
+	client := hcloud.NewClient(hcloud.WithToken(config.HetznerCloudToken))
+	if client == nil {
+		return nil, errors.New("Failed to connect to hetzner cloud")
+	}
+
+	server, _, err := client.Server.Get(context.Background(), config.HetznerCloudServerId)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &HetznerFloatingIpConfigurer{
+		IPConfiguration: ipConfig,
+		client:          client,
+		server:          server,
+		floatingIpId:    config.HetznerCloudIpId,
+		cachedState:     unknown,
+		lastAPICheck:    time.Unix(0, 0),
+		verbose:         config.Verbose,
+	}
+
+	return c, nil
+}
+
+func (c *HetznerFloatingIpConfigurer) queryAddress() bool {
+	if (time.Since(c.lastAPICheck) / time.Second) > 1 {
+		/**We need to recheck the status!
+		 * Don't check too often because of stupid API rate limits
+		 */
+		log.Println("Cached state was too old.")
+		c.cachedState = unknown
+	} else {
+		/** no need to check, we can use "cached" state if set.
+		 * if it is set to UNKNOWN, a check will be done.
+		 */
+		if c.verbose {
+			log.Printf("queryAddress: returning cached status %+v\n", c.cachedState)
+		}
+
+		if c.cachedState == configured {
+			return true
+		} else if c.cachedState == released {
+			return false
+		}
+	}
+
+	floatingIp, _, err := c.client.FloatingIP.Get(context.Background(), c.floatingIpId)
+
+	if c.verbose {
+		log.Printf("queryAddress: state=%v floatingIp=%+v\n", c.cachedState, floatingIp)
+	}
+
+	if err != nil {
+		if c.verbose {
+			log.Printf("queryAddress: err=%v\n", err)
+		}
+
+		c.cachedState = unknown
+		return false
+	} else {
+		c.lastAPICheck = time.Now()
+	}
+
+	if floatingIp.Server != nil && floatingIp.Server.ID == c.server.ID {
+		// We "are" the current failover destination.
+		c.cachedState = configured
+		return true
+	}
+
+	c.cachedState = released
+	return false
+}
+
+func (c *HetznerFloatingIpConfigurer) configureAddress() bool {
+	if c.verbose {
+		log.Printf("configuring floating ip %s on server %s", c.floatingIpId, c.server.Name)
+	}
+
+	floatingIp, _, err := c.client.FloatingIP.Get(context.Background(), c.floatingIpId)
+	if err != nil {
+		log.Printf("failed to query floating ip: %v\n", err)
+		return false
+	}
+
+	action, _, err := c.client.FloatingIP.Assign(context.Background(), floatingIp, c.server)
+	if err != nil {
+		log.Printf("failed to assign floating ip: %v\n", err)
+		return false
+	}
+
+	progressCh, errCh := c.client.Action.WatchProgress(context.Background(), action)
+
+	for {
+		select {
```

Comment:
Can we add a timeout to this select as well? We don't want to wait indefinitely here, or does the hetzner library already catch timeouts and gives us an error after a certain time?

Reply:
Since the `hcloud` library performs HTTP request against their publicly documented API, the request will eventually time out if something is very wrong and the action will send an error on the `errCh`.